scripts/cosine_similarity_preprocessing.py

```diff
@@ -0,0 +1,191 @@
+	
+	print("Reading reviews from each user...")
+	for user in os.listdir(user_data_directory):
+		filename = os.getcwd() + "/../data/yelp_users/" + user
+		readUserReviews(filename)
+
+	restaurant_dict = generate_cosine_similarity_matrix()
+
+	print("initial length of restaurant dictionary: " + str(len(restaurant_dict)))
+	#filter out restaurants that will lead to null chi square values
+	print("Filtering out null chi square value restaurants...")
+	filtered_restaurants = {k:v for (k,v) in restaurant_dict.items()
+		if not is_chi_square_null(v, friend_graph)}
+
+	print("After filtering out null chi square restaurants: " + str(len(filtered_restaurants)))
+
+	print("Generating pickle object...")
+	pickle_object = open("cosSimilarityNoNullRestaurants.pkl", "wb")
+	pickle.dump(filtered_restaurants, pickle_object)
+	pickle_object.close()
+
+	print("Done. Quitting...")
+
+
+def readUserReviews(filename):
+	'''
+	Takes in f, a file handle
+	Reads 5 lines at a time. Line 0 is the restaurant id,
+	line 1 is the content of the review.
+	Updates restaurant dict. to contain rest_id -> (user_id, review)
+	'''
+	global restaurants
+	with open(filename, 'r') as f:
+		#the first line in the file is the user id, don't include it in the sets of 5 lines
+		user_id = f.readline().strip()
+		while True:
+			lines = list(islice(f, 5))
+			if not lines:
+				# EOF reached
+				break
+			restaurant_id = lines[0].strip()
+			review = lines[1].strip()
+			if restaurant_id in restaurants:
+				restaurant_list = restaurants[restaurant_id]
+				restaurant_list.append((user_id, review))
+				restaurants[restaurant_id] = restaurant_list
+			else:
+				restaurants[restaurant_id] = [(user_id, review)]
+
+
+def generate_cosine_similarity_matrix():
+	'''
+	Take in the rest_id -> [(user_id_1, review_1), ...] dictionary.
+	For each restaurant, generate a tfidf matrix of the corpus (all the reviews for that particular restaurant)
+	Use the tfidf matrix to generate a cosine similarity matrix.
+	'''
+	global restaurants
+	global threshold
+	tfidf_vectorizer = TfidfVectorizer()
+	#dictionary {resaurant_id - > {cos_matrix -> cos_matrix, users - > {user_id -> index_into_cos_matrix}}}
+	restaurant_dictionary = {}
+
+	for rest_id, r in restaurants.items():
+		print("Generating dictionary for restaurant " + str(rest_id) + "...")
+		#create the entry in restaurant_dictionary that will be added for this restaurant
+		dict_entry = {}
+		#create the corpus, a list of all the reviews for this restaurant
+		corpus = []
+		#dictionary of user -> index of corresponding row in matrix
+		users = {}
+		row = 0
+
+		#first filter the list of reviews to remove duplicates from the same user
+		reviews = remove_dups(r)
+
+		for review_tuple in reviews:
+			#add the text of the review as a new document in the corpus
+			corpus.append(review_tuple[1])
+			#and add (this user_id -> current row in matrix) to the user dictionary 
+			users[review_tuple[0]] = row
+			row+=1
```

Comment:
#nit 
you could enumerate the reviews and get an index like this
`for i, review_tuple in enumerate(reviews):`

Reply:
What does #nit mean?  That is one of those python things that I find less readable than the way I have it.  wtf python.  I'll change it.

Reply:
oh lmao sorry it's just my way of saying "this is not a super important point and I'm just being nitpicky" lol